### ISSUE TITLE
TypeScript 3.8 Syntax Support

### ISSUE THREAD

**Issue by armano2** [typescript-eslint/typescript-eslint#1436]
3.8 requires new AST nodes and to support the new syntax. 

### Type Import (#1697) (_SUPPORTED_)
```ts
import type { SomeThing } from "./some-module.js";
export type { SomeThing };
```

### `export * as ns` Syntax (#1698) (_SUPPORTED_)
```TS
export * as utilities from "./utilities.js";
```

### ECMAScript Private Fields (_NOT YET SUPPORTED_ - see #3430)

AST: https://github.com/estree/estree/pull/222

**Important Note**
We currently "support" such that no tooling should fail on private fields.

However, the AST that is produced is not considered stable and ***will*** change in future _without a major release_. Any tooling looking to lint private fields should not rely upon the representation (or be prepared to handle a breakage).

**Why don't we have support yet?**

In short: we are awaiting agreement on the AST representation from the ESTree spec.  

For the 3.7 release we jumped the gun and eagerly implemented optional chaining using babel's AST representation, thinking that the ESTree spec, would just go with that. However ESTree is going with a different and incompatible representation entirely. This will cause our project (and users) a lot of pain as we will have to do a _major_, breaking change to the AST to align with what ESLint core will use.

To avoid going through the same pain again with private class fields, we're holding off on official support until ESTree has decided upon a representation.

There are a lot of projects, and a lot of volunteers that collaborate on the ESTree spec, so things take time; please be patient.

```ts
class Person {
    #name: string
    
    constructor(name: string) {
        this.#name = name;
    }

    greet() {
        console.log(`Hello, my name is ${this.#name}!`);
    }
}

let jeremy = new Person("Jeremy Bearimy");

jeremy.#name
```

### Read more:
https://devblogs.microsoft.com/typescript/announcing-typescript-3-8-beta/

**Comment by armano2** [comment]
As for Private Fields babel already supports it as 
```yaml
ClassPrivateProperty
  - key: PrivateName
    - id: Identifier
  - value: Expression
  - static: boolean
  - typeAnnotation: TypeAnnotation
```

**Comment by sebinsua** [comment]
Perhaps a lint rule could be created to only allow types to be imported and exported with the new "type import" syntax? That makes it less likely that Babel users will get types in their JavaScript output.

Is there a feature request issue for that yet?

**Comment by vkrol** [comment]
@sebinsua 
> Perhaps a lint rule could be created to only allow types to be imported and exported with the new "type import" syntax?

See `importsNotUsedAsValues` option https://devblogs.microsoft.com/typescript/announcing-typescript-3-8-beta/#type-only-imports-exports.

_Copy-pasted from https://github.com/babel/babel/issues/8361#issuecomment-577682315._

**Comment by joaovieira** [comment]
This was already mentioned in https://github.com/babel/babel/issues/10981#issuecomment-574333209, but wanted to put it here as well. Would be good i we can use the same node property to identify type imports both from Flow and TS (i.e. `importKind: "type"`), so existing tooling works OOTB with TS. For example, [eslint-plugin-import](https://github.com/benmosher/eslint-plugin-import/search?q=importKind&unscoped_q=importKind) already handles (Flow) type imports in many places.

**Comment by armano2** [comment]
@joaovieira thats how its implemented for now, i'm following up ticket at babel and i will align it if changes will be needed

**Comment by calebeby** [comment]
@vkrol the `importsNotUsedAsValues` TS option only will _tell_ you when you are not using `import type` when you should be. An ESLint rule would be useful to auto-fix it.

**Comment by AlCalzone** [comment]
Just out of interest: When do you guys plan to ship a version with support for TS 3.8?

Type-only imports solve quite a few circularity issues in a project of mine, so I'm eager to use the new features. Unfortunately neither Babel, nor ESLint/Prettier have support yet.

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/pull/1465#issuecomment-591562659

**Comment by jsshapiro** [comment]
While we are waiting for an update, is there any straightforward way to tell eslint to ignore these lines as an interim measure? I already know how to deal with the "unused" complaint, but I don't know how to say "don't apply eslint between here and there" in a source file....

Thanks!

**Comment by bradzacher** [comment]
Nothing should be hard failing right now, as long as you're using the latest version of our tooling.
Private fields won't be lintable yet, but should parse without failing.

Prettier just released v2 which uses our tooling for typescript parsing, and thus supports 3.8

**Comment by jsshapiro** [comment]
So it turns out I had _two_ problems, one of which I failed to diagnose. The issue that initially led me to comment here turned out to be a problem with the vetur plugin in VSCode, which was misdiagnosing the "import type" syntax. I have turned that part of vetur off.

The second problem, which still seems to exist in eslint 6.7.2, is that use of a type name is not recognized as a use. Given an "import type" of a typename T, where T is later used in a type (x:T) rather than as a variable, eslint does not recognize the typename use occurrence as a use of the import, and therefore (incorrectly) claims that the imported name is unused.

With vetur disabled in VSCode, I have been able to suppress the typename complaint from eslint using the unused variable pattern, but it will be really nice when this case is handled and I can take those back out.

It's a hard-running exercise keeping up with the evolution of the language. The eslint team is really impressive.

**Comment by yoyo930021** [comment]
> So it turns out I had _two_ problems, one of which I failed to diagnose. The issue that initially led me to comment here turned out to be a problem with the vetur plugin in VSCode, which was misdiagnosing the "import type" syntax. I have turned that part of vetur off.
> 
> The second problem, which still seems to exist in eslint 6.7.2, is that use of a type name is not recognized as a use. Given an "import type" of a typename T, where T is later used in a type (x:T) rather than as a variable, eslint does not recognize the typename use occurrence as a use of the import, and therefore (incorrectly) claims that the imported name is unused.
> 
> With vetur disabled in VSCode, I have been able to suppress the typename complaint from eslint using the unused variable pattern, but it will be really nice when this case is handled and I can take those back out.
> 
> It's a hard-running exercise keeping up with the evolution of the language. The eslint team is really impressive.

If you try add `Vetur: Use Workspace Dependencies` in config when use vetur ?

**Comment by vegerot** [comment]
Question: Does ESTree still not support private class fields?

**Comment by bradzacher** [comment]
ESTree does not fully support it yet.
It has a preliminary, experimental AST (https://github.com/estree/estree/blob/master/experimental/class-features.md) - however there's no guarantee that this AST will be the final AST until the class feature specs advance to Stage 4.

Looking at the [recent TC39 notes](https://github.com/tc39/notes/blob/master/meetings/2020-09/sept-23.md#status-update-for-class-fields-private-methods-static-class-features) - it looks like the features are very close to proceeding to stage 4.

I'm undecided if we want to move to implement the proposed AST right now. We would be the first implementors, and doing so would break any rules that rely upon the current representation babel uses. There's also no guarantee that the preliminary AST won't change before finalisation - which would be a breaking change to deal with.

I think it's probably better to keep it in limbo for the moment, and keep tracking the TC39 progress.

**Comment by ljharb** [comment]
It’s shipped in multiple browsers - it’s highly unlikely to change in a meaningful way, like every stage 3 feature.

**Comment by dimitropoulos** [comment]
I totally agree about waiting, for whatever it's worth.  Don't want a repeat of the optional chaining situation.

@ljharb the problem is not with whether it's at stage 3 or not - but whether the upstream support exists (in ESTree).  This project can land itself in quite a mess if it implements something that later is changed (which has happened and has happened recently).  ESTree would be a better place to make an issue or offer support.

To me, if the AST structure is firmly decided upon, then it doesn't really matter if it's stage 4.  The thought-experiment you might conduct to verify this would be to ask if anything about this issue would be different if TC-39 had happened to promote it to stage 4 in the last meeting in September.  I think the answer is no: we'd still be right where we are (from the vantage point of `typescript-eslint`, I mean, waiting for ESTree).

**Comment by ljharb** [comment]
@dimitropoulos that's a totally fair point, and is why I continue to hope ESTree changes its policy and starts supporting _shipped_ stage 3 syntax.

**Comment by LukasMachetanz** [comment]
Hey. I am wondering if there is a rough estimation when this will be finished? Thanks in advance. :)

**Comment by bradzacher** [comment]
**When will private fields be ready?**
When the feature reaches stage 4, ESTree's spec will be finalised.
When it's finalised, we can implement it.

**When will that be?**
When it's done. [1]
We don't know and are not at all involved in the timelines or progress of the other projects we are dependent on.

**What's the status of support in this project?**
All 3.8 features are fully supported - apart from private fields.
Private fields have support in that they won't break, and an AST is emitted for them.

**What if I want to lint private fields?**
You can use the experimental AST we produce for private fields.
But just be aware that it is exactly that - experimental, unstable and subject to change without a major bump.

<sup>[1] I can give you this piece of string to measure in the meantime.</sup>

**Comment by LukasMachetanz** [comment]
Hey. I would like to ask again what's the current status? Any updates since https://github.com/typescript-eslint/typescript-eslint/issues/1436#issuecomment-735464951? Thanks.

**Comment by bradzacher** [comment]
We do all of our work in the open. Nothing is private. You can see PRs linked to this issue.
